File: Programmers/Level2/221104/Java/app/src/main/java/NewsClustering.java
public class NewsClustering {
  public int solution(String str1, String str2) {
    int[][] arr1 = new int[27][27];
    int[][] arr2 = new int[27][27];

    String lowerCase = str1.toLowerCase();

    int length = lowerCase.length();
    for (int i = 1; i < length; i += 1) {
      if (lowerCase.charAt(i - 1) >= 'a' &&
          lowerCase.charAt(i - 1) <= 'z' &&
          lowerCase.charAt(i) >= 'a' &&
          lowerCase.charAt(i) <= 'z') {
        arr1[lowerCase.charAt(i - 1) - 97][lowerCase.charAt(i) - 97] += 1;
      }
    }
    lowerCase = str2.toLowerCase();
    length = lowerCase.length();

    for (int i = 1; i < length; i += 1) {
      if (lowerCase.charAt(i - 1) >= 'a'
          && lowerCase.charAt(i - 1) <= 'z'
          && lowerCase.charAt(i) >= 'a'
          && lowerCase.charAt(i) <= 'z') {
        arr2[lowerCase.charAt(i - 1) - 97][lowerCase.charAt(i) - 97] += 1;
      }
    }

    int sum = 0, dif = 0;
    for (int i = 0; i < 27; i += 1) {
      for (int j = 0; j < 27; j += 1) {
        sum += Math.max(arr1[i][j], arr2[i][j]);
        dif += Math.min(arr1[i][j], arr2[i][j]);
      }
    }
    if (sum == 0) return 65536;
    return (int) (65536.0 * dif / sum);
  }
}

File: Programmers/Level2/221104/JavaScript/NewsClustering.js
export const makeArr = (str) => {
  const abc = ['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T', 'U', 'V', 'W', 'X', 'Y', 'Z'];
  str = [...str.toUpperCase()];

  const arr = new Array();

  str.filter((e, i) => {
    if (abc.includes(str[i]) && abc.includes(str[i + 1])) arr.push(str[i] + str[i + 1]);
  });

  return arr;
};

export function solution(str1, str2) {
  const arr1 = makeArr(str1);
  const arr2 = makeArr(str2);

  const set = new Set([...arr1, ...arr2]);
  let union = 0;
  let intersection = 0;

  set.forEach((v) => {
    const tmp1 = arr1.filter((e) => e === v).length;
    const tmp2 = arr2.filter((e) => e === v).length;

    union += Math.max(tmp1, tmp2);
    intersection += Math.min(tmp1, tmp2);
  });

  return union === 0 ? 65536 : Math.floor(intersection / union * 65536);
}
